# Práctica Clase 1 

1) Crear una función capaz de convertir números enteros de base 10 a base 2. Debe recibir como parámetro el número a convertir<br>
Consideraciones:<br> 
a. Tratar de resolverlo sin usar la función format(nro,"b")<br>
b. El pdf "conversion-de-decimal-a-binario.pdf" puede servir de ayuda.


In [6]:
def NumeroBinario(numero):
    '''
    Esta función recibe como parámetro un número entero mayor ó igual a cero y lo devuelve en su 
    representación binaria. Debe recibir y devolver un valor de tipo entero.
    En caso de que el parámetro no sea de tipo entero y mayor a -1 retorna nulo.
    '''
    
    if(type(numero) != int or numero < 0):
        return(None)
    if(numero==0):
        return(0)
    
    sBinario = ""
    while(numero > 0):
        rest = str(numero % 2)
        numero = numero // 2
        sBinario = rest + sBinario        
    return int(sBinario)

In [6]:
lista = ['a','b','c','d','f','g']
pru=lista.reverse

res = ''.join(lista)
print(res)
pru

abcdfg


<function list.reverse()>

In [7]:
#Ejemplo:
numeros = [4,128,35,55,2,1,9,12,156]
for n in numeros:
    print(f"{n} \t => {NumeroBinario(n)}")

4 	 => 100
128 	 => 10000000
35 	 => 100011
55 	 => 110111
2 	 => 10
1 	 => 1
9 	 => 1001
12 	 => 1100
156 	 => 10011100


2) Convertir de decimal a binario las fracciones 1/2, 1/3, 1/4, 1/5, 1/6, 1/7, 1/8, 1/9. Luego analizar los resultados y observar qué particularidad se encuentra en los mismos. Se puede usar Python o una calculadora, lo importante es ver si hay algo que podemos notar...

In [2]:
from math import modf

def FraccionBinaria(decimal,presicion):
    '''
    Esta función recibe como parámetro una fraccion y la presicion, 
    convirtiendo esta fraccion de decimal a . Debe recibir y devolver un valor de tipo entero.
    En caso de que el parámetro no sea de tipo entero y mayor a -1 retorna nulo.
    '''
    decimal, entero = modf(decimal)
    sBinario = ""
    for i in range(presicion):
        decimal = decimal * 2
        decimal, entero = modf(decimal)
        sBinario = sBinario + str(int(entero))        
    return sBinario

In [3]:
#Ejemplo:
#numeros = [1/3,4.25,128.3,0.35,0.55,2.1,1.14,9.1273,0.12,12.156]
numeros = [1/2, 1/3, 1/4, 1/5, 1/6, 1/7, 1/8, 1/9]
numDecimales = 8
for n in numeros:
    decimal, entero = modf(n)
    numero = int(entero)
    print(f"{n:.8f} \t => {NumeroBinario(numero)}.{FraccionBinaria(n,numDecimales)}")

0.50000000 	 => 0.10000000
0.33333333 	 => 0.01010101
0.25000000 	 => 0.01000000
0.20000000 	 => 0.00110011
0.16666667 	 => 0.00101010
0.14285714 	 => 0.00100100
0.12500000 	 => 0.00100000
0.11111111 	 => 0.00011100


Salida esperada:
* Fraccion 1 / 2 :  0.5  y En binario:  0.1
* Fraccion 1 / 3 :  0.3333333333333333  y En binario:  0.010101010101010101010101
* Fraccion 1 / 4 :  0.25  y En binario:  0.01
* Fraccion 1 / 5 :  0.2  y En binario:  0.001100110011001100110011
* Fraccion 1 / 6 :  0.16666666666666666  y En binario:  0.001010101010101010101010
* Fraccion 1 / 7 :  0.14285714285714285  y En binario:  0.001001001001001001001001
* Fraccion 1 / 8 :  0.125  y En binario:  0.001
* Fraccion 1 / 9 :  0.1111111111111111  y En binario:  0.000111000111000111000111